js_modules/dagit/packages/core/src/asset-graph/AssetGraphExplorer.tsx

```diff
@@ -175,15 +175,20 @@ const AssetGraphExplorerWithData: React.FC<WithDataProps> = ({
       let nextOpsNameSelection = token;
 
       if (e.shiftKey || e.metaKey) {
+        // Meta key adds the node you clicked to your existing selection
         let tokensToAdd = [token];
-        if (e.shiftKey && lastSelectedNode && node) {
-          const tokensInRange = opsInRange({
-            graph: assetGraphData,
-            from: lastSelectedNode,
-            to: node,
-          });
-          if (tokensInRange.length) {
-            tokensToAdd = tokensInRange;
+
+        // Shift key adds the nodes between the node you clicked and your existing selection.
+        // To better support clicking a bunch of leaves and extending selection, we try to reach
+        // the new node from each node in your current selection until we find a path.
+        if (e.shiftKey && selectedGraphNodes.length && node) {
+          for (let ii = selectedGraphNodes.length - 1; ii >= 0; ii--) {
+            const from = selectedGraphNodes[ii]!;
+            const tokensInRange = assetKeyTokensInRange({from, to: node, graph: assetGraphData});
+            if (tokensInRange.length) {
+              tokensToAdd = tokensInRange;
+              break;
+            }
```

Comment:
Could do this to avoid the non-null assertion:

```
const reversed = [...selectedGraphNodes].reverse();
for (const from of reversed) {
  ...
}
```

Reply:
oh good call! will switch it out for this 👍